xbmc/pvr/PVRGUIInfo.cpp

```diff
@@ -441,6 +442,16 @@ bool CPVRGUIInfo::GetListItemAndPlayerLabel(const CFileItem *item, const CGUIInf
       // special handling for channels without epg or with radio rds data
       case PLAYER_TITLE:
       case VIDEOPLAYER_TITLE:
+      {
+        const MUSIC_INFO::CMusicInfoTag* tag =
+            CServiceBroker::GetGUI()->GetInfoManager().GetCurrentSongTag();
```

Comment:
This seems to be wrong for two reasons. 

a) PVR should not depend on music specific classes. You even had to insert the include for that.
b) Using Infomanager here creates a circular dependency as this code here is called from the info manager

Reply:
I can get rid of that call to the info manager with this

Reply:
> a) PVR already is related to the music classes through various calls to FillMusicInfoTag()

 ... and I want to get rid of this, not to add more of this.

> b) I can get rid of that call to the info manager and instead use the FileItem music tag 

 ... no dependency to MusicInfoTag in PVR, please.

Reply:
> If including the header in the source a major concern 

 I want to clean up the mess that every component know every other component and every component clss every component, often in circles, which easily lead to deadlocks.

Reply:
Would it be acceptable to just ask for the title string from the FileItem with a new function in FileItem.cpp  No includes would be changed.

I realize I could play a Shoutcast stream through the library to get meta data but I want to keep the great option of recording the radio stream in the PVR

Martin

Reply:
> I realize I could play a Shoutcast stream through the library to get meta data but I want to keep the great option of recording the radio stream in the PVR

@emveepee Really a discussion that belongs in the forum and not in a pull request but please keep in mind that while recording some unencrypted video stream such as MPEG-TS Transport of a ATSC or DVB digital television from example over-the-air Live TV can be considered within fair-use to record, (just like it is for a VHS recorder or a DVD/HDD recorder), I am pretty sure that recording files of Shoutcast or Icecast music streams would not fall under fair-use to record. Because as far as I know  Shoutcast or Icecast music stations are not legally broadcasting music that they paid a royalty to play over the radio, but instead are usually pirate radio stations that simply stream ripped MP3 files as-is, usually not even transcoding them.  It is at least something that Team-Kodi should maybe consider when they normally try to distance themselves from features that could be seen as promoting piracy.

Reply:
@Hedda very good points. Thanks for bringing this up.

Reply:
I respect a user's right to choose which TV streams they can legally watch via IPTV or listen to via Internet radio so let's keep this discussion to PVR.  

@ksooo I want to fix my PR to follow your new standards and it is trivial to to supply the title as a simple string.  Would you accept that change?  I have another change to go with it because right now there is a trivial bug in channel changing radio where the title shows a s a pvr....pvr filename when it shouldn't.

Reply:
We should find a solution where the shoutcast title is not provided by a new CFileItem method, but by the existing CFileItem::GetTitle method and that call has to be in the respective provider (music?, video?, player? whcih is best?), not PVRProvider. This may require some more changes in PVRInfoProvider as it must pass the request on (return false) to the next provider for your use case. Currently it always return true on PLAYER_TITLE and VIDEOPLAYER_TITLE, which means the request is not passed on to the next provider. Passing on the request is the main (and maybe only required change) as other providers already use CFileItem::GetTitle for the gui labels in question, iirc.

Reply:
The problem was there is no existing GetTitle() method but that is what my change was

std::string CFileItem::GetTitle() const
{
  if (HasMusicInfoTag() && !GetMusicInfoTag()->GetTitle().empty())
    return GetMusicInfoTag()->GetTitle();
  if (HasVideoInfoTag() && !GetVideoInfoTag()->m_strTitle.empty())
    return GetVideoInfoTag()->m_strTitle;
  return "";
}

There is also a public m_strFile in Filename but that is used more for the file system and not meta data.

Reply:
What about GetLabel()?

Reply:
PlayerGUIInfo obtains PLAYER_TITLE value from GetLabel()...

Reply:
You are a genius, I hope to know the code as well as you on day.  With this change

      case PLAYER_TITLE:
      case VIDEOPLAYER_TITLE:
        {
          if (channel->IsRadio() && !epgTag)
          {
            return false;
          }

I see the shoutcast data.

Reply:
Not sure skipping everything that has no epg is right here (have to thing about this a moment...), but for sure checking for IsRadio is not needed here.

Reply:
The reason for that was for the user to see the 19055 message if they choose or blank when it falls through.   There already is a small issue since the label duplicates the station name when there is no shoutcast data  but this is an easy fix

if (!epgTag && item->GetLabel() != channel->ChannelName())

perhaps I should just process item->GetLabel() here after that.